### PR TITLE
Add ZoomLimiter to LabelLayer (#504)

### PR DESCRIPTION
**Pull request by Gustl22**
I don't really know if zoom limiter is necessary in `LabelLayer`. Maybe it's good to have. But since using `TileSeparator` the labels aren't displayed twice at tile borders.
And don't know if these few labels, which aren't processed multiple times (at borders) have better performance economization compared to loading the zoom limit tiles of `LabelLayer` (here zoom 18).

Anyway I made `LabelLayer.ZOOM_LIMIT = 18` public, so if set this to `Viewport.MAX_ZOOM_LEVEL` all works similar to before.

### CODE REVIEW

**Comment by devemux86** [comment]
It's not only about labels / symbols at tile edges.
It's also how they are multiplied for each zoom e.g. >17.
If set symbols at polygon parkings, then at large zoom there are multiple symbols per "tiled" polygon.

File: vtm/src/org/oscim/layers/tile/vector/labeling/LabelLayer.java [hunk 1]
```diff
@@ -24,22 +25,28 @@
...
+        ZoomLimiter.IZoomLimiter {
 
     static final Logger log = LoggerFactory.getLogger(LabelLayer.class);
 
     static final String LABEL_DATA = LabelLayer.class.getName();
 
     private static final long MAX_RELABEL_DELAY = 100;
 
+    /* Set to maximum "zoom-min" value for labels in render themes */
+    public static final int ZOOM_LIMIT = 18;
```
Comment (Gustl22) [inline vtm/src/org/oscim/layers/tile/vector/labeling/LabelLayer.java]: Of course have to make it `private` or remove `final` ;)

**Comment by Gustl22** [comment]
> If set symbols at polygon parkings, then at large zooms there are multiple symbols per "tiled" polygon.

Yeah that's a good reason; for other areas, too. 

Unfortunately the zoom limit has to be 18, otherwise it is bothered by render themes.

File: vtm/src/org/oscim/layers/tile/vector/labeling/LabelLayer.java (hunk 1 above)
Comment (devemux86) [reply vtm/src/org/oscim/layers/tile/vector/labeling/LabelLayer.java]: `ZOOM_LIMIT` shouldn't be large or else rendering is similar like now. A sane default could be `17`.

Nevertheless a side effect is that if set in themes `zoom-min` >  `ZOOM_LIMIT` then the labels / symbols are not rendered!
We somehow need to not break themes functionality here.

**Comment by devemux86** [comment]
I merged it for now in [issue_544](https://github.com/mapsforge/vtm/tree/issue_544) branch with mentioned improvements.

Will need more work there to not break themes `zoom-min` experience.

**Comment by Gustl22** [comment]
> Will need more work there to not break themes zoom-min experience.

I didn't work with theme logic so far. I'll have a look at it next week when I have enough time (or can try to solve yourself, if you want)

**Comment by Gustl22** [comment]
I'm afraid. I don't see a proper way to implement that. In `Rule` there's a method [`matchElement`](https://github.com/mapsforge/vtm/blob/master/vtm/src/org/oscim/theme/rule/Rule.java#L75).
But it can't differ for which `RenderStyle` or here the `LabelLayer` it should match. Somehow we have to set equal `LabelLayer` and Symbol + Text + Caption render styles.
So we would need to request Layer information, which isn't available here. We also can't differ, if tile has original zoom level or faked overzoomed level. So it would match for non-overzoomed layers, too.
Even if we manage that, we have to change [`VectorTileLoader`'s](https://github.com/mapsforge/vtm/blob/master/vtm/src/org/oscim/layers/tile/vector/VectorTileLoader.java#L342) `renderText` and `renderSymbol` to call theme hooks on zoom limit tiles if it's overzoomed. But here we don't get these infos, too.
Even then, I didn't manage to display captions, in our example (_I hardcoded the other stuff to test_).
I run out of ideas :worried: 

Main problem is that tiles are independent of Layers, and call information only of its zoom level. So we only could use zoom limits, when nothing in zooms or themes changes in their layers, so it uses exactly what we expect. Tiles can't differ the layers and individually pass information for their needs.

I would suggest, to use zoom limit at 18, where nothing in theme's changes. So symbols and captions aren't divided in 19 and 20.
And can personally increase if want to use higher zooms in themes.
Instead we should fix label positioning, while map creation(?). This would work precisely in any case.

Sorry for that, if you have more ideas, let me know.

**Comment by devemux86** [comment]
> I would suggest, to use zoom limit at 18, where nothing in theme's changes. So symbols and captions aren't divided in 19 and 20.

Default implementation should probably be without zoom limit on themes, that's what users expect.

> Instead we should fix label positioning, while map creation(?). This would work precisely in any case.

I already did that in https://github.com/mapsforge/mapsforge/issues/1061. :slightly_smiling_face: 
Now can use `label-position` safely, either globally as map-writer option or per map element in tag-mapping file.

**Comment by devemux86** [comment]
@Gustl22 thanks for double checking it!

**Comment by Gustl22** [comment]
> Default implementation should probably be without zoom limit on themes, that's what users expect.

So we could set `LabelLayer.ZOOM_LIMIT = Viewport.MAX_ZOOM_LEVEL`, so it would work as there wasn't a ZoomLimiter.

> I already did that in mapsforge/mapsforge#1061.

Very nice. Does it make sense for other polygons/areas, too? Or we individually can check while testing, and change from time to time.

**Comment by devemux86** [comment]
> So we could set LabelLayer.ZOOM_LIMIT = Viewport.MAX_ZOOM_LEVEL, so it would work as there wasn't a ZoomLimiter.

Something like that. In [issue_544](https://github.com/mapsforge/vtm/tree/issue_544) this default value is private and can control the layer via a new constructor.

> Does it make sense for other polygons/areas, too?

What do you mean?
The `label-position` calculation can be applied on all map areas - if set as [map-writer option](https://github.com/mapsforge/mapsforge/blob/master/docs/Getting-Started-Map-Writer.md#advanced-options-only-use-when-you-know-what-you-are-doing).
Or per map element - if set in [tag-mapping](https://github.com/mapsforge/mapsforge/blob/master/mapsforge-map-writer/src/main/config/tag-mapping.xml#L324-L380).

**Comment by Gustl22** [comment]
> What do you mean?

I only meant if it's necessary to change `label-position` [defaults](https://github.com/mapsforge/mapsforge/blob/master/mapsforge-map-writer/src/main/config/tag-mapping.xml), except at buildings, e.g. amenity parking? Especially on large areas, we can't set multiple label positions. But I don't think essential changes are needed.

**Comment by devemux86** [comment]
> if it's necessary to change label-position defaults, except at buildings, e.g. amenity parking

That would be useful, feel free to see what other closed ways - which host labels or symbols - can use the `label-position` in tag-mapping.

I only added it in buildings as the most relevant example for testing.
e.g. lakes usually don't need it in order to allow multiple labels(?)

**Comment by devemux86** [comment]
Thanks!
Merged via 9e52e9e84807a05bf5540996e9cf67a65e620edc, keeping zoom limit to max zoom to allow render themes to work.

**Pull request closed without merging**